Add more details to the SVR Beholder message

cll-gg opened the pull request:
See OEV-172. 

Update SVR TxMessage to include:
* creation time
* chain id
* feed address 



<!--- Does this work have a corresponding ticket?  Please link it here as well as one of:
    - the PR title
    - branch name
    - commit message
[LINK-777](https://smartcontract-it.atlassian.net/browse/LINK-777)
--> 

### Requires

https://github.com/smartcontractkit/chainlink-protos/pull/60

<!--- Does this work depend on other open PRs? Please list them.
- https://github.com/smartcontractkit/chainlink-common/pull/7777777
-->


### Notes

Had to run `modvendor -copy="**/*.a **/*.h" -v` to get the code building, not sure if this is a known issue. (source: https://github.com/marcboeker/go-duckdb/#vendoring)

cll-gg committed:
WIP

cll-gg committed:
Send chain id + current time

cll-gg committed:
Include forwarderDestAddress in Beholder message

cll-gg committed:
Add changeset

cll-gg committed:
go mod tidy

cll-gg committed:
Go mod tidy for all

cll-gg commented:
CodePilot-generated summary:

This pull request includes several updates to the Chainlink codebase, focusing on enhancing the transaction message functionality, updating dependencies, and improving test coverage. Below are the most important changes:

### Enhancements to Transaction Messaging:
* Added fields `CreatedAt`, `ChainId`, and `FeedAddress` to the `TxMessage` struct in `EmitTxMessage` function to send more data with transaction messages (`core/chains/evm/txm/metrics.go`).
* Updated `sendTransactionWithError` to include destination address from metadata when emitting transaction messages (`core/chains/evm/txm/txm.go`).

### Debugging and Logging:
* Added temporary debug print statements to log transaction messages and stored transactions (`core/chains/evm/txm/metrics.go`, `core/chains/evm/txm/storage/inmemory_store.go`). [[1]](diffhunk://#diff-22b0b70beb738f5e805262cb3540be4619c15d347b766717eea680a48035c3dbL98-R111) [[2]](diffhunk://#diff-e2b88623b138204f63c09095de0fcd64bad1c5703d4ec710bc986f487c73c804R154-R155)

### Dependency Updates:
* Updated `chainlink-protos/svr` dependency across multiple `go.mod` files to version `v0.0.0-20250227132844-5581683e4bad`. [[1]](diffhunk://#diff-e62e9f1d5d119cfd293a9429b5ba18709855e60b4c18bf22efbf17ce97527a80L349-R349) [[2]](diffhunk://#diff-16ac5f47ba6841978858e293f48c1b9c75df1ea5aa9fd1171ee857de81510defL361-R361) [[3]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L322-R322) [[4]](diffhunk://#diff-7c016cae50e2281d7a89e537d58b33cb9e01501820ff9301365fc006c94b712bL449-R449) [[5]](diffhunk://#diff-15f8f80fd17ebb70a1da6cbea30cb26566ff9fd8c49137f23869f110c0060b5cL440-R440) [[6]](diffhunk://#diff-1d4e7d3b1a36c04110400d7de579e1b3ec982cdf6ff6922677025b4d98db907eL350-R350) [[7]](diffhunk://#diff-85e1a39b2f2def912a409e0e944fc8f9071dde0051353ffd87c63085a48aed1cL355-R355)

### Test Enhancements:
* Added `destinationAddress` and metadata handling in `TestBackfillTransactions` to improve test coverage (`core/chains/evm/txm/txm_test.go`). [[1]](diffhunk://#diff-5977729fe6f1f405f4a845bc55f57c65cbe99643f3de669f611684e89c4c84a1R228) [[2]](diffhunk://#diff-5977729fe6f1f405f4a845bc55f57c65cbe99643f3de669f611684e89c4c84a1R261-R273) [[3]](diffhunk://#diff-5977729fe6f1f405f4a845bc55f57c65cbe99643f3de669f611684e89c4c84a1R311-R317) [[4]](diffhunk://#diff-5977729fe6f1f405f4a845bc55f57c65cbe99643f3de669f611684e89c4c84a1R326)

### Minor Changes:
* Included the `time` package in `import` statements to support new fields in `TxMessage` (`core/chains/evm/txm/metrics.go`).
* Created a changeset file to document the minor version update due to the new data fields in `TxMessage` (`.changeset/fast-pigs-behave.md`).

dimriou commented on core/chains/evm/txm/txm.go:
```diff
@@ -358,14 +358,25 @@ func (t *Txm) sendTransactionWithError(ctx context.Context, tx *types.Transactio
 	t.metrics.IncrementNumBroadcastedTxs(ctx)
-	if err = t.metrics.EmitTxMessage(ctx, attempt.Hash, address, tx.ToAddress, strconv.FormatUint(*tx.Nonce, 10)); err != nil {
+
+	// get destination address from meta if it exists, to improve txMessage
+	meta, _ := tx.GetMeta()
+	fmt.Printf("Meta is %+v\n", meta)
+
+	destAddress := common.Address{}
+	if meta != nil && meta.FwdrDestAddress != nil {
+		destAddress = *meta.FwdrDestAddress
+	}
+
+	if err = t.metrics.EmitTxMessage(ctx, attempt.Hash, address, tx.ToAddress, destAddress, strconv.FormatUint(*tx.Nonce, 10)); err != nil {
```
Since we now have 3 tx fields, I would instead pass the tx as-is inside the method and unwrap the meta fields there. This way, it is properly abstracted from the txm since it's an svr logic. Also, let's handle the error returned from `GetMeta` since it would be nice to know if it failed during unmarshalling.

cll-gg replied:
Makes sense 👍

cll-gg replied:
Done

dimriou commented on core/chains/evm/txm/metrics.go:
```diff
@@ -95,13 +96,19 @@ func (m *txmMetrics) RecordTimeUntilTxConfirmed(ctx context.Context, duration fl
 		Nonce:       nonce,
+		CreatedAt:   time.Now().UnixMicro(),
+		ChainId:     m.chainID.String(),
+		FeedAddress: forwardedAddress.String(),
```
Maybe keep the `ForwarderAddress` name as the field name, to stay more generic?

cll-gg replied:
Not sure, in the context of SVR we only want to know the feed address, we don't necessarily care that it was forwarded to by the forwarder contract. Or do we?

dimriou replied:
What I meant is this `TxMessage` struct can be used in other use cases besides SVR, where ForwarderAddress doesn't necessarily represent a FeedAddress, hence the renaming. At the same time, this is registered as `svrv1`, so I don't have a strong preference for it, we could keep it as it is.

cll-gg replied:
Indeed, the idea with Beholder is that each domain (product) creates and maintains their own event structs. In case someone (e.g. DF) wants to reuse the TxMessage struct, it'd probably be better to duplicate it into their own domain and adding a protobuf is cheap is anyway.

dimriou commented on core/chains/evm/txm/txm_test.go:
```diff
@@ -254,11 +258,19 @@ func TestBackfillTransactions(t *testing.T) {
 		require.NoError(t, err)
 		txm.metrics = emptyMetrics
 
+		txMeta := &txmgrtypes.TxMeta[common.Address, common.Hash]{
+			FwdrDestAddress: &destinationAddress,
+		}
+		raw, err := json.Marshal(txMeta)
+		require.NoError(t, err)
+		meta := sqlutil.JSON(raw)
+
```
Since we pass an `emptyMetrics` implementation to the TXM and we don't test the Emit functionality, I don't think this makes sense to keep. Maybe add a separate test in `transaction.go` that tests marshalling/unmarhsalling instead? 🤷

cll-gg replied:
I'll have a look, good idea

cll-gg replied:
Done, please check

dimriou commented on core/chains/evm/txm/txm_test.go:
```diff
@@ -296,6 +308,13 @@ func TestBackfillTransactions(t *testing.T) {
 		require.NoError(t, err)
 		txm.metrics = emptyMetrics
 
+		txMeta := &txmgrtypes.TxMeta[common.Address, common.Hash]{
+			FwdrDestAddress: &destinationAddress,
+		}
+		raw, err := json.Marshal(txMeta)
+		require.NoError(t, err)
+		meta := sqlutil.JSON(raw)
+
```
Same as above.

cll-gg committed:
PR feedback: pass full tx, handle error if we can't read meta

cll-gg committed:
Swap arg order to be more consistent

cll-gg committed:
Add test for transaction (specifically for Meta)

cll-gg committed:
go mod tidy

cll-gg committed:
Update chainlink-protos dependency to released version (v1.1.0)

cll-gg committed:
Remove test output and revert updated test and remove not needed test

cll-gg committed:
Remove import

cll-gg committed:
Fix more imports

dimriou commented on core/chains/evm/txm/metrics.go:
```diff
@@ -95,12 +101,22 @@ func (m *txmMetrics) RecordTimeUntilTxConfirmed(ctx context.Context, duration fl
-func (m *txmMetrics) EmitTxMessage(ctx context.Context, tx common.Hash, fromAddress common.Address, toAddress common.Address, nonce string) error {
+func (m *txmMetrics) EmitTxMessage(ctx context.Context, txHash common.Hash, fromAddress common.Address, tx *types.Transaction) error {
+
+	destAddress, err := m.getDestinationAddress(tx)
```
I don't think this is the right approach as anyone, not necessarily SVR, can execute this path. In that case, `Emit` will act as a noop, but we'll see a warning thrown on every transmission. Instead, I would skip the function and write it like this and let the error bubble up, or make it a passthrough if `FwdrDestAddress` is not set:
```suggestion
	meta, err := tx.GetMeta()
	if err != nil {
		return err
	}

	var destAddress string 
	if meta != nil && meta.FwdrDestAddress != nil {
		destAddress = meta.FwdrDestAddress.String()
	}
```

cll-gg replied:
Added these guidelines as discussed on Slack. I'll start a discussion with a wider group to see how we should approach Beholder events for something as wide as TXM, this will be a follow-up item/ticket.

dimriou commented on core/chains/evm/txm/metrics.go:
```diff
@@ -95,12 +101,22 @@ func (m *txmMetrics) RecordTimeUntilTxConfirmed(ctx context.Context, duration fl
-		Nonce:       nonce,
+		ToAddress:   tx.ToAddress.String(),
+		Nonce:       strconv.FormatUint(*tx.Nonce, 10),
+		CreatedAt:   time.Now().UnixMicro(),
```
Just an FYI, this timestamp will be ~0.5s different from the original `attempt.CreatedAt` timestamp, which is stored locally. Perhaps insignificant for our case, but worth noting.

cll-gg replied:
Good to know indeed. I think it's good to keep CreatedAt scoped to the Beholder event for now, in case we want to start monitoring the time between the creation of the tx attempt and the actual sending to chain (or something like it), we can start sending more timestamps.

cll-gg committed:
PR feedback: bubble up error if meta is not set

cll-gg committed:
Small clean up: fix double import

cll-gg committed:
Remove not-needed function

dimriou commented on core/chains/evm/txm/metrics.go:
```diff
@@ -38,13 +42,14 @@ var (
 type txmMetrics struct {
 	metrics.Labeler
 	chainID              *big.Int
+	lggr                 logger.Logger
```
You can drop the logger, it's not needed anymore.

cll-gg replied:
Good point!

cll-gg replied:
Fixed

cll-gg committed:
Remove lggr field, no need

dimriou commented on core/chains/evm/txm/types/transaction_test.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
nit: isn't this marked as an `unused error` by the linter?

cll-gg committed:
No need for keeping a closure reference in a for-loop anymore (https://go.dev/blog/loopvar-preview)

dimriou commented on core/chains/evm/txm/metrics.go:
```diff
@@ -95,12 +98,26 @@ func (m *txmMetrics) RecordTimeUntilTxConfirmed(ctx context.Context, duration fl
 
-func (m *txmMetrics) EmitTxMessage(ctx context.Context, tx common.Hash, fromAddress common.Address, toAddress common.Address, nonce string) error {
+func (m *txmMetrics) EmitTxMessage(ctx context.Context, txHash common.Hash, fromAddress common.Address, tx *types.Transaction) error {
+
```
nit: also this is marked in the linter as `unnecessary leading newline`

cll-gg replied:
Fixed

dimriou commented on core/chains/evm/txm/types/transaction_test.go:
```diff
@@ -0,0 +1,110 @@
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/smartcontractkit/chainlink-common/pkg/sqlutil"
```
nit: same here: error `File is not properly formatted (goimports)`

cll-gg replied:
Fixed

cll-gg committed:
Fix lint issues

cll-gg commented:
Update: updated my local golangci-lint and updated vscode settings to use it properly (https://golangci-lint.run/welcome/integrations/), now linting works locally as well.
